sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,727 @@
+        """Predict which cluster each sample in X belongs to by going down
+        the hierarchical tree in searching of closest leaf cluster.
+
+        In the vector quantization literature, `cluster_centers_` is called
+        the code book and each value returned by `predict` is the index of
+        the closest code in the code book.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            New data to predict.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        labels : ndarray of shape (n_samples,)
+            Index of the cluster each sample belongs to.
+        """
+        check_is_fitted(self)
+
+        X = self._check_test_data(X)
+        x_squared_norms = row_norms(X, squared=True)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+
+        # Note: Well optimized Bisect K-Means should produce clusters with same or
+        # even lower inertia than K-Means.
+        # So after optimization part below may be replaced
+        # with simple '_check_labels_threadpool_limit' function
+
+        # With only one cluster all points have same label
+        if self.n_clusters == 1:
+            return np.zeros(X.shape[0], dtype=np.intc)
+
+        # Get all clusters from inner tree as array
+        all_centers = np.array([val["center"] for val in self._inner_tree.values()])
+
+        # Centers stored in tree have subtracted mean for calculating centroids
+        if not sp.issparse(X):
+            all_centers += self._X_mean
+
+        # Init labels for first two clusters
+        centers_hierarchical = [0, 1]
+        labels = _check_labels_threadpool_limit(
+            X, sample_weight, x_squared_norms, all_centers[:2], self._n_threads
+        )
+
+        # Go down the tree with samples to assign them to proper leaves
+        while len(centers_hierarchical) != self.n_clusters:
+            nested_centers = []
+
+            for idx in centers_hierarchical:
+                center_ids = [idx]
+
+                if self._inner_tree[idx]["children"] is not None:
+                    center_ids = self._inner_tree[idx]["children"]
+                    picked_samples = labels == idx
+
+                    new_labels = _check_labels_threadpool_limit(
+                        X[picked_samples],
+                        sample_weight[picked_samples],
+                        x_squared_norms[picked_samples],
+                        all_centers[center_ids[0] : (center_ids[1] + 1)],
+                        self._n_threads,
+                    )
+
+                    # Label with id of center
+                    new_labels[new_labels == 0] = center_ids[0]
+                    new_labels[new_labels == 1] = center_ids[1]
```

Comment:
Can this fail if `center_ids[0] = 1` ? Then after replacing zeros with `center_ids[0]`, all values in `new_labels` will be 1 ?

Reply:
See that `center_ids` are children of each `centers_hierarchical` which are inited as [0,1] and all nested clusters have always number `last_center_id + 1`.
In that configuration appearing center_ids[0] = 1 would be considered as bug anyway, since none of the centers should have child that is main parent.